load java.instrument module when loading agent

pdbain-ibm opened the pull request:
Java 9 and above need the java.instrument module loaded when
using -javaagent or loading an instrumentation agent via late attach.
Avoid this on legacy versions of Java 9.

Fixes #2030 

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

smlambert commented:
Jenkins test sanity all

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -34,6 +34,8 @@
 static UDATA addToSystemProperty(J9JavaVM * vm, const char * propertyName, const char * segment);
 static UDATA addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader, BOOLEAN enforceJarRestriction);
+#define INSTRUMENT_MODULE "java.instrument"
+UDATA JNICALL loadModule(J9JavaVM *vm, const char *moduleName, jobject *modulePtr);
```
This method should be static in which case JNICALL should be omitted.

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +242,52 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+	if (NULL == jimModules) {
+		rc = CLS_ERROR_NOT_FOUND;
+	} else {
+		jmethodID loadModule =
```
Shouldn't this save `jimModules` in `vm->jimModules`?

gacholio replied:
Not clear what you mean here.  If you're going to cache it, you'll need to move the JVMTI code out into a helper in util (definitely don't duplicate the mutexxed init code),

gacholio replied:
Is there a good reason this isn't just in the VM constant pool?  Could then use findKnownClass to get it.

pdbain-ibm replied:
That I don't know.  I stole the code from jvmtiModules.c.  Maybe @tajila knows.

gacholio replied:
@tajila Actually, VMCP classes are only loaded eagerly if they have fields/methods declared in the constant pool.  If it's just the class, they are loaded on demand.

gacholio replied:
Discussed with Tobi & Peter, we will go with the CP solution to simplify.

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +242,52 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+		if (NULL == loadModule) {
+			rc = CLS_ERROR_NOT_FOUND;
+		} else {
+			jstring moduleNameString =  (*env)->NewStringUTF(env, moduleName);
```
A check that this succeeded should be added. Also please fix the formatting (extra space after `=`).

pdbain-ibm replied:
Thanks.  Fixed.

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +242,52 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+ * @param [in] moduleName name of module to load
+ * @param [out] modulePtr pointer to return module object.  May be null.
+ *
+ * @return CLS_ERROR_NONEE on success, other value on failure
```
Please correct the success code.

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -227,6 +229,9 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
 			if ((*env)->ExceptionCheck(env)) {
 				rc = CLS_ERROR_OUT_OF_MEMORY;
 			}
+			if 	((J2SE_VERSION(vm) >= J2SE_19) && (J2SE_SHAPE(vm) >= J2SE_SHAPE_B165)) {
```
Some wild spacing here.

gacholio replied:
This code should not run if the above code errored.  Need an else or goto cleanup.

pdbain-ibm commented:
I moved the code to load jdk/internal/module/Modules to a utility file per GAC's suggestion
and addressed (I hope) the other comments.  I compiled and did an ad-hoc test.  Could someone run another pull request build?

smlambert commented:
Jenkins test sanity all

keithc-ca commented on runtime/util/moduleclasses.c:
```diff
@@ -0,0 +1,46 @@
+			} else {
+				jclass globalModules = (*env)->NewGlobalRef(env, jimModules);
+				if (NULL == globalModules) {
+					omrthread_monitor_exit(vm->jlmModulesInitMutex);
```
This yields to two calls `omrthread_monitor_exit`.

gacholio replied:
Right, this should just be and else case.

pdbain-ibm replied:
Oops.  Missed that.

gacholio commented on runtime/util/moduleclasses.c:
```diff
@@ -0,0 +1,46 @@
+#include "j9.h"
+
+IDATA
+ensureJimModules(J9JavaVM* vm, JNIEnv* env) {
```
Formatting - also, don't pass both the thread and the VM.  I also despise the use of IDATA as a return value.

gacholio replied:
I suppose IDATA follows existing (if bad) conventions.

pdbain-ibm replied:
... or I could just delete the whole file and put the class in the VM constant pool, as discussed.

pdbain-ibm commented:
@jdekonin or @AdamBrousseau sanity builds are failing: `ERROR: Cannot find label value matching JOB_TYPE:'build' SPEC:'aix_ppc-64_cmprssptrs' and SDK_VERSION:'11'
Finished: FAILURE`  Is this a known issue?

pdbain-ibm commented:
Compiled this locally, but not tested.  @gacholio could you do a preliminary review?

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,71 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
 }
+
+jclass
+getJimModules(J9VMThread *currentThread) {
```
Brace on it's own line,

pdbain-ibm replied:
> Brace on it's own line,
No apostrophe in the possessive of "it".

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,71 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+	} else if (NULL != modulePtr) {
+		*modulePtr = vmModule;
+	}
+	done:
```
No indent.

gacholio replied:
Could do a single assignment to *modulePtr here instead of the two you have.

pdbain-ibm replied:
Sorry, please clarify.

gacholio replied:
I'm saying you can do a single NULL check and assignment at done.

gacholio commented:
Other than a few nitpicks, this looks good to me.

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,71 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+	J9JavaVM *vm = currentThread->javaVM;
+	if (NULL == vm->jimModules) {
+		const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+		/* no need for mutual exclusion since repeating this is harmless */
```
Need to enterVMFromJNI here.

gacholio replied:
So you won't be able to declare and initialize j9JimModules on the same line.

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,71 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+		} else {
+			currentThread->currentException = NULL;
+			currentThread->privateFlags &= ~J9_PRIVATE_FLAGS_REPORT_EXCEPTION_THROW;
+		}
```
...and exitVMToJNI here.

pdbain-ibm replied:
fixed.

gacholio commented:
On further reflection, I'm not sure the constant pool was the right choice, as the module stuff is all being done in a JNI context and used in JNI calls.

Using FindClass is probably a better solution. The mutex can be avoided by doing:

enterVMFromJNI
J9Class *clazz = J9VM_J9CLASS_FROM_JCLASS(vmThread, jclassForModule);
jimModule = &clazz->classObject;
exitVMToJNI

gacholio commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,71 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+				J9VMCONSTANTPOOL_JDKINTERNALMODULEMODULES,
+				J9_FINDKNOWNCLASS_FLAG_INITIALIZE | J9_FINDKNOWNCLASS_FLAG_NON_FATAL);
+		if (NULL != j9JimModules) {
+			vm->jimModules = (jclass)&j9JimModules->classObject;
```
We should add a macro for this later, even though it will look odd to J9CLASS_FROM_JCLASS followed by JCLASS_FROM_J9CLASS.

gacholio replied:
Also probably worth a comment here that the address of j9class->classObject is considered to be a JNI global ref.  It won't keep a class alive, but that's OK here since the class in question is not a candidate for unloading.

pdbain-ibm replied:
https://github.com/eclipse/openj9/issues/2437

gacholio commented:
jenkins test sanity xlinux jdk9

gacholio commented:
jenkins test sanity xlinux jdk9

gacholio commented:
Intentionally cancelled travis build since it will take too long to even start,

pdbain-ibm commented:
@keithc-ca would you kindly review and approve?
Thanks

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,81 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+getJimModules(J9VMThread *currentThread)
+{
+	J9JavaVM *vm = NULL;
+	vm = currentThread->javaVM;
```
nit: this initialization could be collapsed into the declaration.

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,81 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+loadModule(J9JavaVM *vm, const char *moduleName, jobject *modulePtr)
+{
+	UDATA rc = CLS_ERROR_NONE;
+	J9VMThread *currentThread = currentVMThread(vm);
```
Where this function is used, we've already retrieved the current thread: it would make sense to avoid redoing that by passing `currentThread` or `env` instead of `vm`.

pdbain-ibm replied:
Will fix in https://github.com/eclipse/openj9/issues/2437

keithc-ca commented on runtime/vm/classloadersearch.c:
```diff
@@ -237,3 +241,81 @@ addZipToLoader(J9JavaVM * vm, const char * filename, J9ClassLoader * classLoader
+	vmModule = (*env)->CallStaticObjectMethod(env, jimModules, loadModule, moduleNameString);
+	if ((*env)->ExceptionOccurred(env)) {
+		rc = CLS_ERROR_INTERNAL;
+	} else if (NULL != modulePtr) {
```
I like GAC's suggestion to move this after `done` and remove the `*modulePtr = NULL;` above.

pdbain-ibm replied:
Agreed.  Will fix in https://github.com/eclipse/openj9/issues/2437

keithc-ca commented:
Jenkins test sanity win jdk8,jdk10

pdbain-ibm commented:
Per discussion with Keith and GAC, this change is needed ASAP.  Deferring improvements to #2437.

keithc-ca commented:
The windows failures are due to https://github.com/eclipse/openj9/issues/2440.

keithc-ca commented:
Enough builds have passed to satisfy me.
